### PR TITLE
feat(aws-amplify-react-native): Add withOAuth HOC for Cognito Hosted UI

### PR DESCRIPTION
**Pull request by manueliglesias**
*Issue #, if available:*

*Description of changes:*
```ts
// Configure Amplify
Amplify.configure({
  Auth: {
    oauth: {
      // Domain name
      domain: 'some-hosted-ui.auth.xx-xxxx-x.amazoncognito.com',

      // Authorized scopes
      scope: ['phone', 'email', 'profile', 'openid', 'aws.cognito.signin.user.admin'],

      // Callback URL
      redirectSignIn: Linking.makeUrl('/sign-in'),

      // Sign out URL
      redirectSignOut: Linking.makeUrl('/sign-out'),

      // 'code' for Authorization code grant, 
      // 'token' for Implicit grant
      responseType: 'code',

      // optional, for Cognito hosted ui specified options
      options: {
        // Indicates if the data collection is enabled to support Cognito advanced security features. By default, this flag is set to true.
        AdvancedSecurityDataCollectionFlag: true,
      },

      // Optional, showing Expo's WebBrowser usage in ReactNative
      urlOpener: async (url, redirectUrl) => {
        const { type, url: newUrl } = await WebBrowser.openAuthSessionAsync(url, redirectUrl);

        if (type === 'success') {
          await WebBrowser.dismissBrowser();

          if (Platform.OS === 'ios') {
            return Linking.openURL(newUrl);
          }
        }
      }
    }
  }
});

// Build UI
class App extends React.Component {
  render() {
    const {
      oAuthUser: user,
      oAuthError: error,
      hostedUISignIn,
      facebookSignIn,
      googleSignIn,
      amazonSignIn,
      customProviderSignIn,
      signOut
    } = this.props;

    return (
      <SafeAreaView style={styles.safeArea}>
        {user && <Button title="Sign Out" onPress={signOut} icon='logout' />}
        <ScrollView contentContainerStyle={styles.scrollViewContainer}>
          <Text>{JSON.stringify({ user, error, }, null, 2)}</Text>
          {!user && <React.Fragment>
            <Button title="Cognito" onPress={hostedUISignIn} />
            <Button title="Facebook" onPress={facebookSignIn} />
            <Button title="Google" onPress={googleSignIn}  />
            <Button title="Amazon" onPress={amazonSignIn} />
            <Button title="Yahoo" onPress={() => customProviderSignIn('Yahoo')} />
          </React.Fragment>}
        </ScrollView>
      </SafeAreaView>
    );
  }
}

// Create HOC
export default withOAuth(App);
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by manueliglesias** [comment]
Look at the diff with `?w=1` in the url, `Auth.js` shows a lot of changes but it is only because the file used to have wrong line endings.

File: packages/auth/src/Auth.ts [hunk 1]
```diff
@@ -1,1411 +1,1476 @@
...
+
+        try {
+            ({ Buffer } = require('buffer'));
+            ({ Linking } = require('react-native'));
+        } catch (error) { /* Keep webpack happy */ }
+
+        const initialUrl = await Linking.getInitialURL();
+
+        //#region react-native overrides
+        CognitoAuth.prototype.createCORSRequest = (method: string, url: string) => {
```
Comment (undefobj) [inline packages/auth/src/Auth.ts]: Please add a comment linking to this issue: https://github.com/aws/amazon-cognito-auth-js/pull/152
We can remove this function override once this fix is merged into Cognito Auth library, or when we port that library into Amplify and deprecate that library.

File: packages/auth/src/Auth.ts [hunk 2]
```diff
@@ -1,1411 +1,1476 @@
...
+            const xhr = new XMLHttpRequest();
+            xhr.open(method, url, true);
+            return xhr;
+        };
+        (global as any).atob = (global as any).atob
```
Comment (undefobj) [inline packages/auth/src/Auth.ts]: Similar to above, we can remove this if we make Cognito Auth work in RN as well as browser. Please add a comment for future refactors.

File: packages/aws-amplify-react-native/src/Auth/withOAuth.js [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                    break;
+                }
+                case 'signIn_failure':
+                case 'cognitoHostedUI_failure': {
+                    logger.debug('not signed in');
+                    this.setState({ user: null, error: decodeURIComponent(payload.data) });
+                    break;
+                }
+                default:
+                    break;
```
Comment (undefobj) [inline packages/aws-amplify-react-native/src/Auth/withOAuth.js]: :)

File: packages/aws-amplify-react-native/src/Auth/withOAuth.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+export default (Comp, options) => class WithOAuth extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.hostedUISignIn = this.hostedUISignIn.bind(this);
+        this.signOut = this.signOut.bind(this);
+        this._getOAuthUrl = this._getOAuthUrl.bind(this);
+
+        this.urlOpener = (options && options.urlOpener) || defaultUrlOpener;
```
Comment (undefobj) [inline packages/aws-amplify-react-native/src/Auth/withOAuth.js]: Let's add the alternative URL option to the docs

File: packages/auth/src/Auth.ts (hunk 1 above)
Comment (manueliglesias) [reply packages/auth/src/Auth.ts]: Changes in amazon-cognito-auth-js@1.3.2 made this unnecessary.

File: packages/auth/src/Auth.ts (hunk 2 above)
Comment (manueliglesias) [reply packages/auth/src/Auth.ts]: Changes in amazon-cognito-auth-js@1.3.2 made this unnecessary.

File: packages/aws-amplify-react-native/src/Auth/withOAuth.js (hunk 3 above)
Comment (manueliglesias) [reply packages/aws-amplify-react-native/src/Auth/withOAuth.js]: 😝

File: packages/aws-amplify-react-native/src/Auth/withOAuth.js (hunk 4 above)
Comment (manueliglesias) [reply packages/aws-amplify-react-native/src/Auth/withOAuth.js]: Will do!

**Comment by mtraynham** [comment]
Are `buffer` and `react-native` supposed to `peerDependencies` of `@aws-amplify/auth@1.2.17`?

https://github.com/aws-amplify/amplify-js/pull/2665/files#diff-ac40fa61dfab81b51944a4e8ab2de752R59

I don't think this affects anything, but you get a warning now when installing.
```
npm WARN @aws-amplify/auth@1.2.17 requires a peer of buffer@4.x but none is installed. You must install peer dependencies yourself.
npm WARN @aws-amplify/auth@1.2.17 requires a peer of react-native@^0.44.0 but none is installed. You must install peer dependencies yourself.
```

**Pull request merged**